include/RAJA/util/zip_tuple.hpp

```diff
@@ -0,0 +1,352 @@
+  }
+};
+
+struct DeRef
+{
+  template< typename Iter >
+  RAJA_HOST_DEVICE inline auto operator()(Iter&& iter) const
+    -> decltype(*std::forward<Iter>(iter))
+  {
+    return *std::forward<Iter>(iter);
+  }
+};
+
+struct Swap
+{
+  template< typename T0, typename T1 >
+  RAJA_HOST_DEVICE inline int operator()(T0&& t0, T1&& t1) const
+  {
+    using camp::safe_swap;
+    safe_swap(std::forward<T0>(t0), std::forward<T1>(t1));
+    return 1;
+  }
+};
+
+struct IterSwap
+{
+  template< typename T0, typename T1 >
+  RAJA_HOST_DEVICE inline int operator()(T0&& t0, T1&& t1) const
+  {
+    using RAJA::safe_iter_swap;
+    safe_iter_swap(std::forward<T0>(t0), std::forward<T1>(t1));
+    return 1;
+  }
+};
+
+
+/*!
+    \brief Call f on each member of t (f(t)...).
+*/
+template < typename Tuple, typename F, camp::idx_t... Is >
+RAJA_HOST_DEVICE inline
+void zip_for_each_impl(Tuple&& t, F&& f, camp::idx_seq<Is...>)
+{
+  RAJA::detail::sink(std::forward<F>(f)(std::forward<Tuple>(t).template get<Is>())...);
+}
+
+/*!
+    \brief Call f on each member of t0 and t1 (f(t0, t1)...).
+*/
+template < typename Tuple0, typename Tuple1, typename F, camp::idx_t... Is >
+RAJA_HOST_DEVICE inline
+void zip_for_each_impl(Tuple0&& t0, Tuple1&& t1, F&& f, camp::idx_seq<Is...>)
+{
+  RAJA::detail::sink(std::forward<F>(f)(std::forward<Tuple0>(t0).template get<Is>(), std::forward<Tuple1>(t1).template get<Is>())...);
+}
+
+/*!
+    \brief Call f on each member of t (f(t)...).
+*/
+template < typename Tuple, typename F >
+RAJA_HOST_DEVICE inline
+void zip_for_each(Tuple&& t, F&& f)
+{
+  zip_for_each_impl(std::forward<Tuple>(t), std::forward<F>(f), typename camp::decay<Tuple>::IdxSeq{});
+}
+
+/*!
+    \brief Call f on each member of t0 and t1 (f(t0, t1)...).
+*/
+template < typename Tuple0, typename Tuple1, typename F >
+RAJA_HOST_DEVICE inline
+void zip_for_each(Tuple0&& t0, Tuple1&& t1, F&& f)
+{
+  static_assert(std::is_same<typename camp::decay<Tuple0>::IdxSeq, typename camp::decay<Tuple1>::IdxSeq>::value,
+      "Tuple0 and Tuple1 must have the same size");
+  zip_for_each_impl(std::forward<Tuple0>(t0), std::forward<Tuple1>(t1), std::forward<F>(f), typename camp::decay<Tuple0>::IdxSeq{});
+}
+
+
+/*!
+    \brief Tuple used by ZipIterator for storing multiple references and values.
+    Acts like a reference to its members allowing copy/move construction/assignment
+    based on the reference type of the zip_tuple.
+*/
+template < bool is_val, typename ... Ts >
+struct zip_tuple
+{
+  using value_type = RAJA::tuple<Ts...>;
+
+  // zip_tuple type with opposite is_val
+  template < typename ... Os >
+  using opp_tuple = zip_tuple<!is_val, Os...>;
+
+  // camp::idx_seq for this type, also used by zip_for_each
+  using IdxSeq = camp::make_idx_seq_t<sizeof...(Ts)>;
+
+  // constructor from types convertible to Ts
+  template < typename ... Os
+           , typename = concepts::enable_if<DefineConcept(concepts::convertible_to<Ts>(camp::val<Os&&>()))...> >
```

Comment:
Concept checking should be wrapped within a requires_ -- else the compiler error gets a bit insane (fails on template substitution rather than overload resolution). The detector idiom doesn't work correctly otherwise. Overload resolution is a much cleaner diagnostic.

```cpp
typename = concepts::enable_if<concepts::requires<DefineConcept(...)>>
```

Reply:
This is quite true, also the concepts checking in the current master on CAMP is not super reliable, I have an active branch for re-working this but it needs to be partially redone to bring it back to C++11 compliance so we can use it over here.  If this is something you need here, I’ll prioritize getting the updated base detection machinery and such in there so this will work the way you want.

Reply:
Should say, it was quite reliable for a long time, but something happened a while ago that caused certain conditions to sneak in that don’t actually cause substitution failure, so some conditions succeed when they shouldn’t on some not terribly well-conforming compilers.

Reply:
Ok, I'll change that.
Regarding your question @trws do you think this is a reasonable path to follow in the first place? See my recent comment on how this differs from normal/expected semantics.

Reply:
Actually, I'm not sure where I copied that usage of DefineConcept from, is there a preferred way of doing this?

Reply:
@MrBurmark, you were defining the concept fine. It's just when you check a concept, we need it to be in a `requires_`

Reply:
I copied how it was done in camp, take a look.
If this looks reasonable, should this be added to camp?